class-fields-registry.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+final class Fields_Registry {
+	protected $datastores = [];
+
+	protected $datastore_types = [];
+
+	protected $control_types = [];
+
+	protected static $instance = null;
+	protected        $controls = [];
+
+	/**
+	 * @return $this
+	 */
+	public static function get_instance() : Fields_Registry {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new self;
+		}
+
+		return self::$instance;
+	}
+
+	public function register_datastore_type( string $type, string $instance ) {
+		if ( ! isset( $this->datastore_types[ $type ] ) ) {
+			$this->datastore_types[ $type ] = $instance;
+		}
+
+		return $this;
+	}
+
+	public function register_control_type( string $type, string $instance ) {
+		if ( ! isset( $this->control_types[ $type ] ) ) {
+			$this->control_types[ $type ] = $instance;
+		}
+
+		return $this;
+	}
+
+	/**
+	 * @param string $type
+	 *
+	 * @return string|null
+	 */
+	protected function get_field_control_type( string $type ) {
+		return isset( $this->control_types[ $type ] ) && is_string( $this->control_types[ $type ] ) && is_subclass_of( $this->control_types[ $type ], Field_Control::class ) ? $this->control_types[ $type ] : null;
+	}
+
+	/**
+	 * @param string $type
+	 *
+	 * @return string|null
+	 */
+	protected function get_field_datastore_type( string $type ) {
+		return isset( $this->datastore_types[ $type ] ) && is_string( $this->datastore_types[ $type ] ) && is_subclass_of( $this->datastore_types[ $type ], Field_Data_Store::class ) ? $this->datastore_types[ $type ] : null;
+	}
+
+	public function register_json( string $file ) {
+		$fields = json_decode( file_get_contents( $file ), true );
+
+		foreach ( $fields as $field ) {
```

Comment:
Should this logic be abstracted into a function that can be called if someone doesn't want to use JSON but wants to use the same structure via PHP? So maybe a `register_fields` method?